File: realtime/nodejs/auto_test_asr_v2.js
const WebSocket = require('ws');
const fs = require('fs');
const crypto = require('crypto');
const path = require('path');
const { program } = require('commander');

// 常量定义
const TIME_PER_CHUNK = 0.2;  // 每次发送的音频数据的时间长度，单位：s
const NUM_CHANNEL = 1;  // 声道数
const NUM_QUANTIFY = 16;  // 量化位数
const SAMPLE_RATE = 16000;  // 采样频率
const BYTES_PER_CHUNK = Math.floor(SAMPLE_RATE * NUM_QUANTIFY * TIME_PER_CHUNK * NUM_CHANNEL / 8);
const SLEEP_TIME_DURATION = 100; // 100ms, 转换为毫秒

// 生成签名
function generateSignature(appId, apiKey) {
    const ts = Math.floor(Date.now() / 1000).toString();
    const md5 = crypto.createHash('md5');
    md5.update(appId + ts);
    const baseString = md5.digest();
    
    const hmac = crypto.createHmac('sha1', apiKey);
    hmac.update(baseString);
    const signa = hmac.digest('base64');
    
    return { signa, ts };
}

// 发送音频数据
async function sendAudioData(ws) {
    return new Promise((resolve, reject) => {
        const filename = path.join(__dirname, "../dataset/test.wav");
        const readStream = fs.createReadStream(filename, {
            highWaterMark: BYTES_PER_CHUNK
        });

        const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

        readStream.on('data', async chunk => {
            const startTime = Date.now();
            
            if (ws.readyState === WebSocket.OPEN) {
                ws.send(chunk);
                const elapsed = Date.now() - startTime;
                if (elapsed < SLEEP_TIME_DURATION) {
                    await sleep(SLEEP_TIME_DURATION - elapsed);
                }
            }
        });

        readStream.on('end', () => {
            if (ws.readyState === WebSocket.OPEN) {
                ws.send('');  // 发送空字符串表示结束
            }
            resolve();
        });

        readStream.on('error', reject);
    });
}

// 接收识别结果
function receiveRecognitionResult(ws, printMode) {
    ws.on('message', data => {
        if (!data) return;
        
        try {
            const asrJson = JSON.parse(data);
            const isFinal = asrJson.is_final || false;
            const segId = asrJson.seg_id || 0;
            const asr = asrJson.asr || "";
            const type = asrJson.type || "";

            if (printMode === "typewriter") {
                if (type === "asr") {
                    if (isFinal) {
                        process.stdout.write(`\r${segId}:${asr}\n`);
                    } else {
                        process.stdout.write(`\r${segId}:${asr}`);
                    }
                }
            } else {
                if (isFinal) {
                    console.warn(asrJson);
                } else {
                    console.log(asrJson);
                }
            }
        } catch (err) {
            console.error('Error parsing message:', err);
        }
    });
}

async function connectToServer(printMode, asrType, recall) {
    const appId = "test1";
    const appSecret = "2258ACC4-199B-4DCB-B6F3-C2485C63E85A";
    const baseUrl = "ws://192.168.2.141:2001/asr-realtime/v2/ws";
    
    const { signa, ts } = generateSignature(appId, appSecret);
    const url = `${baseUrl}?appid=${appId}&ts=${ts}&signa=${encodeURIComponent(signa)}&asr_type=${asrType}&trans_mode=0&target_lang=ru&pd=court&recall=${recall}`;

    const ws = new WebSocket(url);

    return new Promise((resolve, reject) => {
        ws.on('open', async () => {
            console.log('Connected to server');
            receiveRecognitionResult(ws, printMode);
            
            try {
                await sendAudioData(ws);
            } catch (err) {
                console.error('Error sending audio:', err);
                ws.close();
            }
        });

        ws.on('error', err => {
            console.error('WebSocket error:', err);
            reject(err);
        });

        ws.on('close', () => {
            console.log('Connection closed');
            resolve();
        });
    });
}

// 命令行参数解析
program
    .option('--mode <type>', 'Output mode: typewriter or json', 'typewriter')
    .option('--asr_type <type>', 'ASR recognition mode: sentence or word', 'word')
    .option('--recall', 'Enable recall mode', true)
    .parse(process.argv);

const options = program.opts();

// 主程序执行
connectToServer(options.mode, options.asr_type, options.recall)
    .catch(err => {
        console.error('Program error:', err);
        process.exit(1);
    });

File: realtime/python/auto_test_asr_v2.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import asyncio
import websockets
from websockets.client import WebSocketClientProtocol
import os, sys, hmac, hashlib, time
from urllib.parse import quote
import orjson,json
import random, base64
from loguru import logger
import argparse
from dotenv import load_dotenv
load_dotenv()

time_per_chunk = 0.2  # 每次发送的音频数据的时间长度，单位：s
num_channel = 1  # 声道数
num_quantify = 16  # 量化位数
sample_rate = 16000  # 采样频率
bytes_per_chunk = int(sample_rate * num_quantify * time_per_chunk * num_channel / 8)
sleep_time_duration = 0.1 # 每次发送音频数据后等待的时间长度，单位：s

sys.path.append(os.path.dirname(os.path.abspath(__file__)))
sys.path.append(os.path.join(os.path.dirname(__file__), "../dataset/asr"))

async def send_audio_data(websocket: WebSocketClientProtocol, audio_file: str):
    try:
        filename = audio_file
        chunk_size = bytes_per_chunk
        
        with open(filename, "rb") as file:
            while True:
                sleep_time = time.time()
                data = file.read(chunk_size)
                if not data:
                    break
                await websocket.send(data)
                if (time.time() - sleep_time) < sleep_time_duration:
                    await asyncio.sleep(abs(sleep_time_duration - (time.time() - sleep_time)))
                sleep_time = time.time()
                await asyncio.sleep(0)
        await websocket.send('')
        #await websocket.send('{"end": true}')
        await asyncio.sleep(120)
    except (websockets.exceptions.ConnectionClosedError, KeyboardInterrupt) as e:
        logger.info(f"Connection closed or interrupted: {e}")
    except Exception as e:
        logger.error(f"Error in send_audio_data: {e}")

async def receive_recognition_result(websocket: WebSocketClientProtocol, print_mode: str):
    try:
        while True:
            result = await websocket.recv()  # 接收语音识别结果
            if not result:
                break
            #print("Recognition result:", result)
            asr_json = orjson.loads(result)
            is_final = asr_json.get("is_final", False)
            seg_id = asr_json.get("seg_id", 0)
            asr = asr_json.get("asr", "")
            type = asr_json.get("type", "")
            if print_mode == "typewriter":
                if type == "asr":
                    # 打字机效果输出：根据 is_final 决定是否换行
                    if is_final:
                        print(f"\r{seg_id}:{asr}", flush=True)
                    else:
                        print(f"\r{seg_id}:{asr}", end="", flush=True)
            else:
                if is_final:
                    logger.warning(f"{asr_json}")
                else:
                    logger.info(f"{asr_json}")
        await asyncio.sleep(120)  # 释放控制权，允许其他任务执行
    except Exception as err:
        print(f"receive_recognition_result error: {repr(err)}")

def generate_signature(app_id: str, api_key: str) -> str:
    """
    @param app_id: 应用程序ID
    @param api_key: 应用程序秘钥
    @return: 签名, 时间戳
    """
    ts: str = str(int(time.time()))
    tt = (app_id + ts).encode('utf-8')
    md5 = hashlib.md5()
    md5.update(tt)
    baseString = md5.hexdigest()
    baseString = bytes(baseString, encoding='utf-8')

    apiKey = api_key.encode('utf-8')
    signa = hmac.new(apiKey, baseString, hashlib.sha1).digest()
    signa = base64.b64encode(signa)
    signa = str(signa, 'utf-8')
    return signa, ts

async def connect_to_server(print_mode: str, asr_type: str, audio_file: str):
    app_id = os.getenv("ZMEET_APP_ID")
    app_secret = os.getenv("ZMEET_APP_SECRET")
    
    if not app_id or not app_secret:
        raise ValueError("缺少必需的环境变量：ZMEET_APP_ID 或 ZMEET_APP_SECRET 未设置")
    
    base_url = "wss://audio.abcpen.com:8443/asr-realtime/v2/ws"
    signa, ts = generate_signature(app_id, app_secret)
    
    # 更新 URL，添加声纹识别参数
    url = (f"{base_url}?appid={app_id}&ts={ts}&signa={quote(signa)}"
           f"&asr_type={asr_type}&voiceprint={args.voiceprint}"
           f"&voiceprint_org_id={args.voiceprint_org_id}"
           f"&voiceprint_tag_id={args.voiceprint_tag_id}")
    
    try:
        async with websockets.connect(url) as websocket:
            task_send = asyncio.create_task(send_audio_data(websocket, audio_file))
            task_receive = asyncio.create_task(receive_recognition_result(websocket, print_mode))

            done, pending = await asyncio.wait(
                [task_receive, task_send], 
                return_when=asyncio.FIRST_COMPLETED
            )
            
            # 取消所有未完成的任务
            for task in pending:
                task.cancel()
            # 等待取消的任务完成
            await asyncio.gather(*pending, return_exceptions=True)
    except Exception as e:
        logger.error(f"Connection error: {e}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='ASR Client')
    parser.add_argument('--mode', 
                       choices=['typewriter', 'json'],
                       default='typewriter',
                       help='Output mode: typewriter (default) or json')
    parser.add_argument('--asr_type',
                       choices=['sentence', 'word'],
                       default='word',
                       help='ASR recognition mode: sentence (default) or word')
    parser.add_argument('--audio_file',
                       type=str,
                       default=os.path.join(os.path.dirname(__file__), "../dataset/asr/1006_20241223_081645_full_audio.wav"),
                       help='Path to the audio file (default: ../dataset/asr/3-1-60s.wav)')
    parser.add_argument('--voiceprint',
                       type=str,
                       default='1',
                       help='Enable voiceprint recognition (default: 1)')
    parser.add_argument('--voiceprint_org_id',
                       type=str,
                       default=None,  # 将使用 app_id 作为默认值
                       help='Organization ID for voiceprint (default: same as app_id)')
    parser.add_argument('--voiceprint_tag_id',
                       type=str,
                       default=None,  # 将使用 app_id 作为默认值
                       help='Tag ID for voiceprint (default: same as app_id)')

    args = parser.parse_args()
    
    # 如果未指定 org_id 和 tag_id，使用 app_id
    app_id = os.getenv("ZMEET_APP_ID")
    if args.voiceprint_org_id is None:
        args.voiceprint_org_id = app_id
    if args.voiceprint_tag_id is None:
        args.voiceprint_tag_id = app_id

    asyncio.run(connect_to_server(args.mode, args.asr_type, args.audio_file))
